### PR TITLE
allow no limit for buffered stream combinators

### LINKED ISSUE #801: BufferUnordered without size limit
As `BufferUnordered` is wrapping `FuturesUnordered` now, we can expose a similar infinite-size buffering combinator. This is actually possible now if you specify a large-enough limit (`usize::MAX`), but it looks like a hack.

What I'm trying to do is to replace `spawn` with a more tree-like model, so we have a large state machine instead of cluttering many types of futures into the event loop.

### LINKED ISSUE #2189: Add option for unbounded `buffer_unordered`
Would it be possible to add an option to `buffer_unordered` to not supply a limit? Why is it possible to provide `None` in `for_each_concurrent` and not in `buffer_unordered`?

### PR DESCRIPTION
**Pull request by ibraheemdev**
Resolves #2189, Resolves #801

### CODE REVIEW

File: futures-util/src/stream/stream/mod.rs [hunk 1]
```diff
@@ -1285,6 +1285,13 @@ pub trait StreamExt: Stream {
     /// buffered at any point in time, and less than `n` may also be buffered
     /// depending on the state of each future.
     ///
+    /// The limit argument is of type `Into<Option<usize>>`, and so can be
+    /// provided as either `None`, `Some(10)`, or just `10`. Note: a limit of zero is
```
[the remark this answers is not in the document]
Comment (taiki-e) [reply futures-util/src/stream/stream/mod.rs:1289]: Hmm... after reading [past discussions on for_each_concurrent's limit](https://github.com/rust-lang/futures-rs/pull/1194#discussion_r209455196), I'm starting to feel it's ok to ignore limit=0.

[the remark this answers is not in the document]
Comment (taiki-e) [reply futures-util/src/stream/stream/mod.rs:1289]: Yes, it's okay to remove those tests.

(FWIW, the current master's buffered's behavior has not been backported to 0.3, and passing zero to a 0.3 buffered basically not work well. https://github.com/rust-lang/futures-rs/issues/2348)

File: futures-util/src/stream/stream/mod.rs [hunk 2]
```diff
@@ -1285,6 +1285,13 @@ pub trait StreamExt: Stream {
...
+    /// interpreted as no limit at all, and will have the same result as passing in `None`.
+    ///
+    /// This method is only available when the `std` or `alloc` feature of this
+    /// library is activated, and it is activated by default.
```
Comment (taiki-e) [inline futures-util/src/stream/stream/mod.rs:1293]: This seems to be a duplicate of L1297-L1298.

**Comment by ibraheemdev** [comment]
I updated the branch, but this pull request doesn't seem to be reflecting the changes...

**Comment by ibraheemdev** [comment]
I updated the branch, but this pull request doesn't seem to be reflecting the changes...

**Review by taiki-e — COMMENTED** [review]
LGTM. Could you remove `# Panics` sections from `buffered`/`buffer_unordered` docs?

https://github.com/rust-lang/futures-rs/blob/a32ba3f3229257b193ad65cc5d746f830cc8ffa4/futures-util/src/stream/stream/mod.rs#L1297
https://github.com/rust-lang/futures-rs/blob/a32ba3f3229257b193ad65cc5d746f830cc8ffa4/futures-util/src/stream/stream/mod.rs#L1327

**Review by taiki-e — APPROVED** [review]
Thanks!

**Pull request merged**